src/fs.rs

```diff
@@ -97,7 +112,217 @@ fn write_directory_contents(dir: &Directory, fd: &OwnedFd, repo: &Repository) ->
     Ok(())
 }
 
+// NB: hardlinks not supported
 pub fn write_to_path(repo: &Repository, dir: &Directory, output_dir: &Path) -> Result<()> {
     let fd = openat(CWD, output_dir, OFlags::PATH | OFlags::DIRECTORY, 0.into())?;
     write_directory_contents(dir, &fd, repo)
 }
+
+pub struct FilesystemReader<'repo> {
+    st_dev: u64,
+    repo: Option<&'repo Repository>,
+    inodes: HashMap<u64, Rc<Leaf>>,
+    root_mtime: i64,
+}
+
+impl<'repo> FilesystemReader<'repo> {
+    fn read_xattrs(&mut self, fd: &OwnedFd) -> Result<BTreeMap<Box<OsStr>, Box<[u8]>>> {
+        // flistxattr() and fgetxattr() don't with with O_PATH fds, so go via /proc/self/fd. Note:
+        // we want the symlink-following version of this call, which produces the correct behaviour
+        // even when trying to read xattrs from symlinks themselves.  See
+        // https://gist.github.com/allisonkarlitskaya/7a80f2ebb3314d80f45c653a1ba0e398
+        let filename = proc_self_fd(fd);
+
+        let mut xattrs = BTreeMap::new();
+
+        let names_size = listxattr(&filename, &mut [])?;
+        let mut names = vec![0; names_size];
+        let actual_names_size = listxattr(&filename, &mut names)?;
+        ensure!(
+            actual_names_size == names.len(),
+            "xattrs changed during read"
+        );
+
+        let names: Vec<u8> = names.into_iter().map(|c| c as u8).collect(); // fml
```

Comment:
Then we could drop this

Reply:
I'd love to drop it, but your above suggestion, for example, gets me:

```
mismatched types
   --> src/fs.rs:140:54
    |
140 |         let actual_names_size = listxattr(&filename, &mut names)?;
    |                                 ---------            ^^^^^^^^^^ expected `&mut [i8]`, found `&mut Vec<u8>`
    |                                 |
    |                                 arguments to this function are incorrect
    |
    = note: expected mutable reference `&mut [i8]`
               found mutable reference `&mut Vec<u8>`
note: function defined here
   --> /var/home/lis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustix-0.38.40/src/fs/xattr.rs:140:8
    |
140 | pub fn listxattr<P: path::Arg>(path: P, list: &mut [c::c_char]) -> io::Result<usize> {
    |        ^^^^^^^^^


```

Reply:
We could also transmute, but I've been absolutely loathe to write the word `unsafe`... there are only a very small number of counterexamples in this entire repository.